ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java

```diff
@@ -136,6 +138,40 @@ public boolean setReplicaSealedState(ReplicaId replicaId, boolean isSealed) {
     }
   }
 
+  @Override
+  public boolean setReplicaStoppedState(List<ReplicaId> replicaIds, boolean isStopped) {
+    Set<String> replicasToUpdate = new HashSet<>();
+    for (ReplicaId replicaId : replicaIds) {
+      if (!(replicaId instanceof AmbryReplica)) {
+        throw new IllegalArgumentException(
+            "HelixParticipant only works with the AmbryReplica implementation of ReplicaId");
+      }
+      replicasToUpdate.add(replicaId.getPartitionId().toPathString());
+    }
+    boolean setResult;
+    synchronized (helixAdministrationLock) {
+      logger.trace("Getting stopped replicas from instanceConfig");
+      List<String> stoppedList = getStoppedReplicas();
+      if (stoppedList == null) {
+        stoppedList = new ArrayList<>();
+      }
+      Set<String> stoppedReplicas = new HashSet<>(stoppedList);
+      if (!isStopped) {
+        replicasToUpdate.retainAll(stoppedReplicas);
+        logger.info("Removing replicas {} from stopped list", Arrays.toString(replicasToUpdate.toArray()));
+        stoppedList.removeAll(replicasToUpdate);
+        setResult = setStoppedReplicas(stoppedList);
+      } else {
+        stoppedReplicas.retainAll(replicasToUpdate);
+        replicasToUpdate.removeAll(stoppedReplicas);
```

Comment:
What you want is UNION (for markStop) and SUBTRACTION (for unmarkStop). For sets, these are `addAll()` and `removeAll()`, respectively. So I think this would suffice:
```java
if (markStopped) { // add to STOPPED list:
  stoppedSet.addAll(replicasToUpdateSet); // union is basically addAll() for sets.
  setResult = setStoppedReplicas(Arrays.asList(stoppedSet));
} else { // remove from STOPPED list:
  stoppedList.removeAll(replicasToUpdateSet);
  setResult = setStoppedReplicas(stoppedList);
}
```

Reply:
Thanks for pointing it out. Actually, I am making changes on this piece code and might use the `replicasToUpdate` after `removeAll()` to determine if we need to invoke `setStoppedReplicas`. When `replicasToUpdate.isEmpty()`, it's unnecessary to update the instanceConfig in Helix again. (I will find a chance to discuss it with you offline)